x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmConfig.java

```diff
@@ -27,9 +27,24 @@ public RealmConfig(RealmIdentifier identifier, Settings settings, Environment en
         this.identifier = identifier;
         this.settings = settings;
         this.env = env;
+        this.threadContext = threadContext;
         this.enabled = getSetting(RealmSettings.ENABLED_SETTING);
+        if (hasSetting(RealmSettings.ORDER_SETTING.apply(type())) == false) {
+            throw new IllegalArgumentException("'order' is a mandatory parameter for realm config. " +
+                "Found invalid realm config: '" + identifier.name + "'\n" +
+                "Please see the breaking changes documentation."
+            );
+        }
         this.order = getSetting(RealmSettings.ORDER_SETTING);
+    }
+
+    public RealmConfig(RealmIdentifier identifier, Settings settings, Environment env, ThreadContext threadContext, int order) {
+        this.identifier = identifier;
+        this.settings = settings;
+        this.env = env;
         this.threadContext = threadContext;
+        this.enabled = getSetting(RealmSettings.ENABLED_SETTING);
+        this.order = order;
```

Comment:
The new constructor is mostly used by test code, but it is also used by production code, see [Realms](https://github.com/elastic/elasticsearch/pull/51195/files#diff-ff593e0560017b528a889b09ee2de166R315) and [ReservedRealm](https://github.com/elastic/elasticsearch/pull/51195/files#diff-13b2f866e83efa400fc0999b3a6834a6L66).

I also think adding a new constructor is not ideal due to some of the concerns you mention. It however seems to be a better choice comparing to the other choice, i.e. Add `order` into the `settings` parameter on caller side so that it can just keep using the existing constructor. 

This approach should work fine for tests. But in production code, the `settings` object is referenced in many other classes, e.g. `Environment`, `Realms`, etc. So when adding FileRealm (in [Realms.java](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Realms.java#L309)), if we modify the `settings` object to add `order`, we will need to do something like the follows:

```java
var newSettings = Settings.Builder().put(settings).build();
... new RealmConfig(..., newSetings, ...), ...;
```
So the `newSettings` is now passed into RealmConfig which will store a reference to it. Now we have two `Settings` object across the application. They are mostly identical, only differs in the `order` parameter. And this is just FileRealm, the same applies to native realm, for which we will have a third `settings`. I personally think this is worse than adding a new constructor. That is why the new constructor is added so we can avoid manipulating the settings object.

With above being said, I realise it is making an assumption that the logic for mandating the `order` setting resides in the `RealmConfig` class. If we can move away from this assumption, like you suggested, enforce the logic in `Realms.initRealms`, we can probably make the change set much smaller. 

Initially I thought about adding the logic into `Security` but that was too *high* level. So we decided to add it in `RealmConfig` itself, where it is *low* level the logic is close to its natural owner. But maybe we need something in between. `Realms` could be the right candidate especially it is where the order uniqueness is enforced.

Reply:
My choice would also be to only inspect the `order` setting in`Realms#initRealms` (where the `order` value duplication is checked). That's because the `order` is only really used in that place, specifying the **initial** order in the realm chain (which would then change as we reorder the chain during authentications), so it's not suitable as a final local variable.
This way we would also avoid having to specify the `order` parameter for the "default" realms (`file` and `native`) which is simply ignored anyways.

Your approach works as well and I can't argue strongly against it, so this LGTM to me even though I would've approached it differently.